io/include/traccc/io/event_data.hpp

```diff
@@ -0,0 +1,96 @@
+ * (c) 2024 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+// Project include(s).
+#include "traccc/edm/cell.hpp"
+#include "traccc/edm/cluster.hpp"
+#include "traccc/edm/measurement.hpp"
+#include "traccc/edm/particle.hpp"
+#include "traccc/edm/spacepoint.hpp"
+#include "traccc/edm/track_candidate.hpp"
+#include "traccc/geometry/silicon_detector_description.hpp"
+#include "traccc/io/csv/cell.hpp"
+#include "traccc/io/csv/hit.hpp"
+#include "traccc/io/csv/measurement.hpp"
+#include "traccc/io/csv/measurement_hit_id.hpp"
+#include "traccc/io/csv/particle.hpp"
+#include "traccc/utils/seed_generator.hpp"
+
+// Detray include(s).
+#include "detray/core/detector.hpp"
+#include "detray/io/frontend/detector_reader.hpp"
+
+// Vecmem include(s).
+#include <vecmem/memory/memory_resource.hpp>
+
+// System include(s).
+#include <map>
+#include <string>
+
+namespace traccc {
+
+struct event_data {
+
+    public:
+    // Type definitions
+    using detector_type = detray::detector<detray::default_metadata,
+                                           detray::host_container_types>;
+
+    event_data() = delete;
+
+    event_data(const std::string& event_dir, const std::size_t event_id,
+               vecmem::memory_resource& resource, const detector_type& det,
+               const silicon_detector_description::host* dd);
+
+    void fill_cca_result(
+        const cluster_container_types::host& clusters,
+        const measurement_collection_types::host& measurements);
+
+    track_candidate_container_types::host generate_truth_candidates(
+        seed_generator<detector_type>& sg, vecmem::memory_resource& resource);
+
+    // IO EDM containers
+    std::vector<io::csv::cell> m_cells;
+    std::vector<io::csv::hit> m_hits;
+    std::vector<io::csv::measurement> m_measurements;
+    std::vector<io::csv::measurement_hit_id> m_meas_hit_ids;
+    std::vector<io::csv::particle> m_particles;
+
+    std::map<cell, particle> m_cell_to_particle_map;
+    std::map<particle_id, particle> m_particle_map;
+    // Measurement to the contributing particle map
+    std::map<measurement, std::map<particle, std::size_t>> m_meas_to_ptc_map;
+    std::map<measurement, std::map<particle, std::size_t>>
+        m_truth_meas_to_ptc_map;
+    // CCA measurement to the contributing particle map
+    std::map<measurement, std::map<particle, std::size_t>>
+        m_found_meas_to_ptc_map;
+    // Particle to its Measurements map
+    std::map<particle, std::vector<measurement>> m_ptc_to_meas_map;
+    // Measurement to its track parameter map
+    std::map<measurement, std::pair<point3, point3>> m_meas_to_param_map;
+    // Modules per event
+    cell_module_collection_types::host m_modules;
+    // cluster to truth measurement map
+    std::map<vecmem::vector<cell>, measurement> m_cluster_to_truth_meas_map;
+    // CCA measurement to cluster map
+    std::map<vecmem::vector<cell>, measurement> m_cluster_to_found_meas_map;
```

Comment:
I am getting **very** worried about all of these different maps. :frowning: Especially ones keyed on vectors of cells. That seems like the most fragile thing ever. Where cell order changes could fully upset the logic.

Also, what about `m_modules`? That thing will be removed very-very soon. The reconstruction algorithms are already not using it. Why use it here? (Searching through the changes, I don't see `m_modules` used **anywhere**. Which is at least good, you should be able to completely remove it.)

In fact I'm pretty sure that some of these maps are not currently used by any client code. Please look that through, and only keep the ones that are actually needed. Because this interface looks very unmaintainable right now. :frowning:

Reply:
OK. I removed the maps that are not being used in this PR.